apps/opik-backend/src/main/java/com/comet/opik/domain/ChatCompletionService.java

```diff
@@ -16,170 +14,139 @@
-                    throw new ClientErrorException(openAiHttpException.getMessage(), openAiHttpException.code());
+            llmProviderClient.getLlmProviderError(runtimeException).ifPresent(llmProviderError -> {
+                if (Response.Status.Family.familyOf(llmProviderError.code()) == Response.Status.Family.CLIENT_ERROR) {
+                    throw new ClientErrorException(llmProviderError.message(), llmProviderError.code());
                 }
-                throw new ServerErrorException(openAiHttpException.getMessage(), openAiHttpException.code());
-            }
+
+                throw new ServerErrorException(runtimeException.getMessage(), llmProviderError.code());
+            });
+
             throw new InternalServerErrorException(UNEXPECTED_ERROR_CALLING_LLM_PROVIDER);
         }
+
         log.info("Created chat completions, workspaceId '{}', model '{}'", workspaceId, request.model());
         return chatCompletionResponse;
     }
 
     public ChunkedOutput<String> createAndStreamResponse(
             @NonNull ChatCompletionRequest request, @NonNull String workspaceId) {
         log.info("Creating and streaming chat completions, workspaceId '{}', model '{}'", workspaceId, request.model());
-        var openAiClient = getAndConfigureOpenAiClient(request, workspaceId);
         var chunkedOutput = new ChunkedOutput<String>(String.class, "\r\n");
-        openAiClient.chatCompletion(request)
-                .onPartialResponse(chatCompletionResponse -> send(chatCompletionResponse, chunkedOutput))
-                .onComplete(() -> close(chunkedOutput))
-                .onError(throwable -> handle(throwable, chunkedOutput))
-                .execute();
-        log.info("Created and streaming chat completions, workspaceId '{}', model '{}'", workspaceId, request.model());
+
+        try {
+            var llmProviderClient = llmProviderFactory.getService(workspaceId, request.model());
+
+            llmProviderClient.generateStream(
+                    request,
+                    workspaceId,
+                    getMessageHandler(chunkedOutput),
+                    getCloseHandler(chunkedOutput),
+                    getErrorHandler(chunkedOutput, llmProviderClient));
+            log.info("Created and streaming chat completions, workspaceId '{}', model '{}'", workspaceId,
+                    request.model());
+        } catch (BadRequestException exception) {
+            handleChunckedException(chunkedOutput, new ErrorMessage(HttpStatus.SC_BAD_REQUEST, exception.getMessage()));
+        }
+
         return chunkedOutput;
     }
 
-    private OpenAiClient getAndConfigureOpenAiClient(ChatCompletionRequest request, String workspaceId) {
-        var llmProvider = getLlmProvider(request.model());
-        var encryptedApiKey = getEncryptedApiKey(workspaceId, llmProvider);
-        return newOpenAiClient(encryptedApiKey);
+    private RetryUtils.RetryPolicy newRetryPolicy() {
```

Comment:
This can be built in a module

Reply:
I agree, however please note that this is strictly bound to `LlmProviderClientConfig` and therefore won't be reused outside this context at the moment, therefore I don't see the point of doing it right now. This could of course be changed in the future if we decide to reuse this logic somewhere.

Reply:
Yeah, but splitting the use case and business logic from the infra-related code would be beneficial.

Reply:
Agreed to fix as part of #987